shell/platform/windows/flutter_windows_engine.cc

```diff
@@ -796,4 +795,39 @@ void FlutterWindowsEngine::OnApplicationLifecycleEnabled() {
   lifecycle_manager_->BeginProcessingClose();
 }
 
+void FlutterWindowsEngine::OnWindowStateEvent(HWND hwnd,
+                                              WindowStateEvent event) {
+  lifecycle_manager_->OnWindowStateEvent(hwnd, event);
+}
+
+void FlutterWindowsEngine::ProcessExternalWindowMessage(HWND hwnd,
+                                                        UINT message,
+                                                        WPARAM wparam,
+                                                        LPARAM lparam) {
+  if (lifecycle_manager_) {
+    std::optional<flutter::WindowStateEvent> event = std::nullopt;
+
+    switch (message) {
+      case WM_SHOWWINDOW:
+        event = wparam ? flutter::WindowStateEvent::kShow
+                       : flutter::WindowStateEvent::kHide;
+        break;
+      case WM_SIZE:
+        event = wparam == SIZE_MINIMIZED ? flutter::WindowStateEvent::kHide
+                                         : flutter::WindowStateEvent::kShow;
+        break;
+      case WM_SETFOCUS:
+        event = flutter::WindowStateEvent::kFocus;
+        break;
+      case WM_KILLFOCUS:
+        event = flutter::WindowStateEvent::kUnfocus;
+        break;
+    }
```

Comment:
This message translation logic is very similar to the logic in `WindowsLifecycleManager`. Can we push this translation logic down to `WindowsLifecycleManager`?

Also, why is this logic to translate external window messages to `WindowStateEvent`s different from the translation logic for Flutter's  windows messages? Shouldn't they be the same?

Reply:
They cannot be the same, because the flutter window never expects to receive `WM_SHOWWINDOW`, and when it receives `WM_SIZE`, it does not indicate in `wparam` the change in window size as is usually the case.

Reply:
> the flutter window never expects to receive `WM_SHOWWINDOW`, and when it receives `WM_SIZE`, it does not indicate in `wparam` the change in window size as is usually the ca

Are you referring to the messages received by `Window::HandleMessage`? I was referring to the logic for the top-level windows that are now in `WindowsLifecycleManager::WindowProc` and `WindowsLifecycleManager::ExternalWindowMessage`, I would expect these to mirror each other however there's differences. For example, on `WM_SHOWWINDOW` with a `TRUE` `wparam`:

* `WindowsLifecycleManager::WindowProc` doesn't do anything
* `WindowsLifecycleManager::ExternalWindowMessage` creates a `WindowStateEvent::kShow` event

Reply:
Sorry, I don't fully understand the differences. Here's my understanding of the current logic:

Scenario | [Flutter's top-level window](https://github.com/yaakovschectman/flutter-engine/blob/bad630200785a8e02c9a05b7cb4e6ec04ee94d68/shell/platform/windows/windows_lifecycle_manager.cc#L41-L98) | [External top-level window](https://github.com/yaakovschectman/flutter-engine/blob/bad630200785a8e02c9a05b7cb4e6ec04ee94d68/shell/platform/windows/windows_lifecycle_manager.cc#L245-L271)
-- | -- | --
`WM_CLOSE` | If last window, requests application quit | Ignored
`WM_SIZE` | If `wparam` indicates window was maximized or restored, generates "show" window state event<br><br>If `wparam` indicates window was minimized, generates "hide" window state event | If `wparam` indicates window was minimized, generates "hide" window state event. Otherwise, generates "show" window state event
`WM_SHOWWINDOW` | If `wparam` is `FALSE`, generates "hide" window state event | If `wparam` is `FALSE`, generates "hide" window state event.<br><br>If `wparam` is `TRUE`, generates "show" window event
`WM_DESTROY` | Generates "hide" window state event | Ignored
`WM_SETFOCUS` | Ignored | Generates "focus" window state event
`WM_KILLFOCUS` | Ignored | Generates "unfocus" window state event

The differences in `WM_SETFOCUS` and `WM_KILLFOCUS` makes sense, we handle this message using the Flutter's "inner" window.

However, I'm still curious about these differences:

1. If the external top-level window receives a `WM_CLOSE` message, why don't we request application exist if it is the last window?
2. If the external top-level window receives a `WM_CLOSE` message with `wparam` of `SIZE_MAXHIDE` or `SIZE_MAXSHOW`, why does that generate a "show" window state event?
3. If the Flutter top-level window receives a `WM_SHOWWINDOW` with a `wparam` of `TRUE`, why don't we generate a "show" window state event? We do this if the external top-level window receives such a message, would it be better to be consistent?
4. Why don't we generate a "hide" window message if an external top-level receives a `WM_DESTROY` message?

If these differences are intentional, please add comments to explain.

Reply:
1.) The logic for `WM_CLOSE` was not part of this PR and was in place beforehand. I believe it is the case that it does not send the request to the framework to check if it should cancel quitting; it will just process the message as usual and close if appropriate. The app quit request worked not by explicitly quitting when allowed to do so, but by forwarding the close message to the rest of the WndProc logic, which would usually end up at the default WndProc and end up closing the program.
2.) Do you mean `WM_SIZE`? If so, the messages may be redundant, but to my understanding, they should at least not end up inaccurate.
3.) That is probably fine to generate the show event as well.
4.) Same as 3.

Reply:
Does the recent commit address these concerns any?

Reply:
> 1.) The logic for `WM_CLOSE` was not part of this PR and was in place beforehand. I believe it is the case that it does not send the request to the framework to check if it should cancel quitting; it will just process the message as usual and close if appropriate. The app quit request worked not by explicitly quitting when allowed to do so, but by forwarding the close message to the rest of the WndProc logic, which would usually end up at the default WndProc and end up closing the program.

I talked to @gspencergoog about this scenario, and we're thinking that the Dart app should be notified of app exit if the last window is an "external" non-Flutter window. As a result, the Windows embedder will need to be able to consume an external window's `WM_CLOSE`. I'm fine with punting this work until a follow-up PR, but, we should update `FlutterEngine::ProcessExternalWindowMessage` and `FlutterDesktopEngineProcessExternalWindowMessage` to allow consuming messages (ideally they should mirror `FlutterViewController::HandleTopLevelWindowProc` and `FlutterDesktopViewControllerHandleTopLevelWindowProc`).

> 2.) Do you mean `WM_SIZE`? If so, the messages may be redundant, but to my understanding, they should at least not end up inaccurate.

Ah yes, I meant `WM_SIZE` 👍 

> Does the recent commit address these concerns any?

The new implementation looks good modulo the new external window `WM_CLOSE` comment above. Thanks for updating!

Reply:
This API call now returns a `bool` indicating whether or not the handler should continue processing the message. For now, it always returns `false`, i.e. the message was not consumed, continue to process it.

Reply:
Should we mirror the top-level Flutter window message handler's API? The [C++ API returns a `std::optional<LRESULT>`](https://github.com/flutter/engine/blob/182e1189873bbec33d96b59f20d3b365e7aaf9ba/shell/platform/windows/client_wrapper/include/flutter/flutter_view_controller.h#L50-L58), the [C API has an `LRESULT* result` out parameter](https://github.com/flutter/engine/blob/182e1189873bbec33d96b59f20d3b365e7aaf9ba/shell/platform/windows/public/flutter_windows.h#L119-L125).

Reply:
I think it still should return a `bool` to indicate whether the message still needs to be processed, but I can an add an `LRESULT*`.

Reply:
For the C++ API, could we use the std::optional<LRESULT> approach there too? If it has a `LRESULT` value, it's the equivalent of returning `true` with the out parameter set. If it doesn't have a value, it's the equivalent of returning `false`. The benefits of the `std::optional` approach:

1. It aligns with the `FlutterViewController` API
2. It's more idiomatic for C++